### DISCUSSION TITLE
Suggestion regarding PR assignees and approvals

### DISCUSSION THREAD

**Discussion by WorldLanguages** [ScratchAddons/ScratchAddons#5238] [Discussion]
"Org(anization) member" means a user that is member of the @ScratchAddons organization. As of October 21st 2022, there's 11 people in the organization: https://github.com/orgs/ScratchAddons/people

Note that this suggestion might possibly not work well in practice, so I would not suggest setting assignees in every one of the 50+ PRs that are currently open. The PR approval rules below do apply since today, tho.
This suggestion only applies to this repository.

# Regarding PR approvals
- PRs require 2 approvals to be merged (rule remains unchanged).
- If the PR author is an org member, it counts as 1 approval.
- Approvals from GitHub users outside the organization may count (use common sense). Users outside of the organization that review PRs frequently may be proposed to join the organization as a member.
- In cases where the 2 approval threshold is not officially reached, but any of the 2 above cases applied, an admin must be pinged/mentioned so that the PR can be force-merged. This is because GitHub isn't aware of approvals from non-org members, and a PR author can't self-approve. **UPDATE:** you can now ping all admins through `@ScratchAddons/admins` 

# Regarding PR assignees
- Pull requests, unless backlogged, should be assigned to 1 or more organization member(s). There may be many assignees in a single PR, but 1 is enough.
- If the PR author is an org member, they may self-assign their own PR.
- The assignee should request review from other org members or contributors, when they see fit. If a PR has 2 assignees, those 2 approvals are enough to merge.
- The assignee decides in what moment in time the PR should be merged. Org members should not merge PRs if they're not the assignee, even if it reached 2 approvals (there can be exceptions, use common sense).
- The assignee should take over the pull request if the original PR author goes inactive.
- The assignee is in charge of setting the milestone for the PR.
- Org members should self-assign. Org members shouldn't assign other members. If a PR doesn't have an assignee after some time, an org member aware of that should ping/mention possible assignees. In practice, I'd expect the first org member that sees a new PR has been created pinging/mentioning possible assignees.
- An org member may self-unassign if they want.
- Personally, I'd like to be the assignee in every "new addon" PR, and most scope:core PRs too. Org members may decide their own preferences, and ideally make them public like I just did.

# Relevant pages for org members
- (New) "What's happening" board: https://github.com/orgs/ScratchAddons/projects/1
- Review requests: https://github.com/pulls/review-requested. **If you don't plan to review a PR any time soon, please remove the review request.** If you want to be automatically requested-review when specific file(s) change, submit a PR to change the .github/CODEOWNERS file.
- Assigned PRs: https://github.com/pulls/assigned
- Check GitHub notifications for @mentions in any issues or PRs. If possible, read every GH notification from the org, but if not, pay close attention to the mentions.

# Goals of this suggestion
- Giving an actual meaning to review requests. The CODEOWNERS file has been recently changed to reflect this.
- Letting every PR have an "expert", instead of every org member having a rough knowledge about every open PR.
- Org members should know what to prioritize: 1. assigned PRs 2. requested reviews 3. general discussion in issues listed on "What's happening" board
- If you have read this, react to this discussion with a thumbs up emoji.
- Offering more predictability and better communication for PR authors that are outside the GitHub organization and haven't joined the development Discord server.
- Moving discussion outside of the Discord server, into dedicated GitHub issues.

**Comment by WorldLanguages** [comment]
Pinging @mxmou (other org members have already been pinged on Discord)

**Reply by cobaltt7** [reply]
not all of them :eyes:

**Comment by WorldLanguages** [comment]
Please reply to this comment if there's any feedback about the new "What's happening" board.

**Comment by cobaltt7** [comment]
> * Approvals from GitHub users outside the organization may count (use common sense).

problem is that then only admins can merge that pr

> * If possible, read every GH notification from the org, but if not, pay close attention to the mentions.

FYI: i do, just usually not for a day or two

> * Org members may decide their own preferences, and ideally make them public like I just did.

tbh I'm really not that active in sa anymore, i review prs that look interesting to me (mainly bc i don't want to have to figure out how to test it again lol shh)

**Reply by WorldLanguages** [reply]
> problem is that then only admins can merge that pr

This is considered. See the sentence that starts with `In cases where the 2 approval threshold is not officially reached`.

**Comment by apple502j** [comment]
I feel like this makes us like a bureaucratic system bound by all sorts of rules. I review PRs whenever I feel like it, and does not review PRs unless I support its addition. (e.g. I don't think necropost checker belongs, so I'm not reviewing it) - and when I think it's a good addition I review, test, give approval, wait for someone to do second approval, and merge.

Having to take care old PR also doesn't sound great. If a PR is inactive for months without the first approval, it should either be reopened by different authors or just get straight up closed, since it's probably not useful. If it does get first approval we can spam ping on Discord asking for another one.

**Reply by WorldLanguages** [reply]
> a bureaucratic system bound by all sorts of rules

If you're not the assignee for any PR, the only new rule is that you should not press the merge button. Most new rules only apply to the assignee. The only other recommendation is to remove review requests that you don't plan to review any time soon.

> I review PRs whenever I feel like it, and do not review PRs unless I support its addition.

As I said above, PR review requests aren't changing much, they just get a bit more meaning vs the one they used to have. But still, no pressure around pending review requests, unless it's been multiple months. You are encouraged to reject review rejects explicitly, instead of keeping them pending forever.

> (e.g. I don't think necropost checker belongs, so I'm not reviewing it)

If you think a new addon submission shouldn't belong on the addon library, you should comment that on the PR/issue.

> when I think it's a good addition I review, test, give approval, wait for someone to do second approval, and merge.

You can continue to do that, but you shouldn't merge unless you're the assignee, or unless you're 100% sure that the PR is ready to merge.

> Having to take care old PR also doesn't sound great. If a PR is inactive for months without the first approval, it should either be reopened by different authors or just get straight up closed, since it's probably not useful.

Every PR is different and has a different context. If a PR only requires little changes to get merged, such as waitForElement optimizations, addon.json changes, etc. and the PR author is inactive, it doesn't make sense to open a new PR just for that. In other cases, it's better to close the PR. It depends in every case, and the assignee can always self-unassign when wanted.

> If it does get first approval we can spam ping on Discord asking for another one.

We can still do that, but one of the goals here is to use GitHub functionality instead of Discord.
